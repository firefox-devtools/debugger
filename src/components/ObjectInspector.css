.arrow svg {
  fill: var(--theme-splitter-color);
  margin-top: 3px;
  transition: transform 0.25s ease;
  width: 10px;
}

html:not([dir="rtl"]) .arrow svg {
  margin-right: 5px;
  transform: rotate(-90deg);
}

html[dir="rtl"] .arrow svg {
  margin-left: 5px;
  transform: rotate(90deg);
}

/* TODO (Amit): html is just for specificity. keep it like this? */
html .arrow.expanded svg {
  transform: rotate(0deg);
}

.arrow.hidden {
  visibility: hidden;
}
