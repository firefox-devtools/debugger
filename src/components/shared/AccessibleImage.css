.img {
  /* default height an width which will likely be overrode */
  width: 12px;
  height: 12px;
  /* makes span appear like an image */
  display: inline-block;
}

.img.arrow.arrow.expanded {
  transform: rotate(0deg);
}
