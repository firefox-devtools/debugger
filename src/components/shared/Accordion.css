:root {
  --accordion-header-background: var(--theme-toolbar-background);
}

:root.theme-dark {
  --accordion-header-background: #141416;
}

.accordion {
  background-color: var(--theme-sidebar-background);
  width: 100%;
}

.accordion ._header {
  background-color: var(--accordion-header-background);
  border-bottom: 1px solid var(--theme-splitter-color);
  display: flex;
  font-size: 12px;
  padding: 4px;
  transition: all 0.25s ease;
  width: 100%;
  height: 24px;
  align-items: center;

  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -o-user-select: none;
  user-select: none;
  cursor: default;
}

.accordion ._header:hover {
  background-color: var(--theme-toolbar-background-hover);
}

.accordion ._header button svg,
.accordion ._header:hover button svg {
  fill: currentColor;
  height: 16px;
}

.accordion ._content {
  border-bottom: 1px solid var(--theme-splitter-color);
  font-size: 12px;
}

.accordion div:last-child ._content {
  border-bottom: none;
}

.accordion ._header .header-buttons {
  display: flex;
  margin-left: auto;
  padding-right: 5px;
}

.accordion .header-buttons .add-button {
  font-size: 180%;
  text-align: center;
  line-height: 16px;
}

.accordion .header-buttons button {
  color: var(--theme-body-color);
  border: none;
  background: none;
  outline: 0;
  padding: 0;
  width: 16px;
  height: 16px;
}

.accordion .header-buttons button::-moz-focus-inner {
  border: none;
}
