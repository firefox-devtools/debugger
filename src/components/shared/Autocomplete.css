
.autocomplete {
  flex: 1;
  width: 100%;
}

.autocomplete .no-result-msg {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  color: var(--theme-graphs-full-red);
  font-size: 24px;
  padding: 4px;
  word-break: break-all;
}

.autocomplete .no-result-msg .sad-face {
  width: 24px;
  margin: 0 4px;
  line-height: 0;
  flex-shrink: 0;
}

.autocomplete .no-result-msg .sad-face svg {
  fill: var(--theme-graphs-full-red);
}
