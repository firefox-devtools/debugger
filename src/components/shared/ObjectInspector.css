.object-value .unavailable {
  color: var(--theme-comment);
}
