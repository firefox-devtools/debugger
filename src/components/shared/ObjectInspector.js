// @flow
import { DOM as dom, PropTypes, createFactory, Component } from "react";
import classnames from "classnames";
import Svg from "./Svg";
import Rep from "./Rep";
import previewFunction from "./previewFunction";
import { MODE } from "devtools-reps";
import {
  nodeIsOptimizedOut,
  nodeIsMissingArguments,
  nodeHasProperties,
  nodeIsFunction,
  nodeIsPrimitive,
  isDefault,
  getChildren,
  createNode
} from "../../utils/object-inspector";

import _ManagedTree from "./ManagedTree";
const ManagedTree = createFactory(_ManagedTree);

export type ObjectInspectorItemContentsValue = {
  actor: string,
  class: string,
  displayClass: string,
  extensible: boolean,
  frozen: boolean,
  ownPropertyLength: number,
  preview: Object,
  sealed: boolean,
  type: string
};

type ObjectInspectorItemContents = {
  value: ObjectInspectorItemContentsValue
};

type ObjectInspectorItem = {
  contents: Array<ObjectInspectorItem> & ObjectInspectorItemContents,
  name: string,
  path: string
};

type DefaultProps = {
  onLabelClick: (
    item: ObjectInspectorItem,
    params: {
      depth: number,
      focused: boolean,
      expanded: boolean,
      setExpanded: () => any
    }
  ) => void,
  onDoubleClick: (
    item: ObjectInspectorItem,
    params: {
      depth: number,
      focused: boolean,
      expanded: boolean
    }
  ) => void,
  autoExpandDepth: number
};

// This implements a component that renders an interactive inspector
// for looking at JavaScript objects. It expects descriptions of
// objects from the protocol, and will dynamically fetch child
// properties as objects are expanded.
//
// If you want to inspect a single object, pass the name and the
// protocol descriptor of it:
//
//  ObjectInspector({
//    name: "foo",
//    desc: { writable: true, ..., { value: { actor: "1", ... }}},
//    ...
//  })
//
// If you want multiple top-level objects (like scopes), you can pass
// an array of manually constructed nodes as `roots`:
//
//  ObjectInspector({
//    roots: [{ name: ... }, ...],
//    ...
//  });

// There are 3 types of nodes: a simple node with a children array, an
// object that has properties that should be children when they are
// fetched, and a primitive value that should be displayed with no
// children.

class ObjectInspector extends Component {
  static defaultProps: DefaultProps;
  actors: any;

  constructor() {
    super();

    this.actors = {};

    const self: any = this;
    self.getChildren = this.getChildren.bind(this);
    self.renderItem = this.renderItem.bind(this);
  }

  getChildren(item: ObjectInspectorItem) {
    const { getObjectProperties } = this.props;
    const { actors } = this;

    return getChildren({
      getObjectProperties,
      actors,
      item
    });
  }

  renderItem(
    item: ObjectInspectorItem,
    depth: number,
    focused: boolean,
    _: Object,
    expanded: boolean,
    { setExpanded }: () => any
  ) {
    let objectValue;
    let label = item.name;
    if (nodeIsOptimizedOut(item)) {
      objectValue = dom.span({ className: "unavailable" }, "(optimized away)");
    } else if (nodeIsMissingArguments(item)) {
      objectValue = dom.span({ className: "unavailable" }, "(unavailable)");
    } else if (nodeIsFunction(item)) {
      objectValue = null;
      label = previewFunction(item);
    } else if (nodeHasProperties(item) || nodeIsPrimitive(item)) {
      const object = item.contents.value;
      objectValue = Rep({ object, mode: MODE.TINY });
    }

    return dom.div(
      {
        className: classnames("node object-node", {
          focused,
          "default-property": isDefault(item)
        }),
        style: { marginLeft: depth * 15 },
        onClick: e => {
          e.stopPropagation();
          setExpanded(item, !expanded);
        },
        onDoubleClick: event => {
          event.stopPropagation();
          this.props.onDoubleClick(item, {
            depth,
            focused,
            expanded
          });
        }
      },
      Svg("arrow", {
        className: classnames({
          expanded: expanded,
          hidden: nodeIsPrimitive(item)
        })
      }),
      dom.span(
        {
          className: "object-label",
          dir: "ltr",
          onClick: event => {
            event.stopPropagation();
            this.props.onLabelClick(item, {
              depth,
              focused,
              expanded,
              setExpanded
            });
          }
        },
        label
      ),
      dom.span({ className: "object-delimiter" }, objectValue ? ": " : ""),
      dom.span({ className: "object-value" }, objectValue || "")
    );
  }

  render() {
    const { name, desc, loadObjectProperties, autoExpandDepth } = this.props;

    let roots = this.props.roots;
    if (!roots) {
      roots = [createNode(name, name, desc)];
    }

    return ManagedTree({
      itemHeight: 20,
      getParent: item => null,
      getChildren: this.getChildren,
      getRoots: () => roots,
      getKey: item => item.path,
      autoExpand: 0,
      autoExpandDepth,
      autoExpandAll: false,
      disabledFocus: true,
      onExpand: item => {
        if (nodeHasProperties(item)) {
          loadObjectProperties(item.contents.value);
        }
      },
      renderItem: this.renderItem
    });
  }
}

ObjectInspector.displayName = "ObjectInspector";

ObjectInspector.propTypes = {
  autoExpandDepth: PropTypes.number,
  name: PropTypes.string,
  desc: PropTypes.object,
  roots: PropTypes.array,
  getObjectProperties: PropTypes.func.isRequired,
  loadObjectProperties: PropTypes.func.isRequired,
  onLabelClick: PropTypes.func.isRequired,
  onDoubleClick: PropTypes.func.isRequired
};

ObjectInspector.defaultProps = {
  onLabelClick: () => {},
  onDoubleClick: () => {},
  autoExpandDepth: 1
};

export default ObjectInspector;
