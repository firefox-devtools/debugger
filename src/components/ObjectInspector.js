const React = require("react");
const classnames = require("classnames");
const ManagedTree = React.createFactory(require("./utils/ManagedTree"));
const Svg = require("./utils/Svg");
const Rep = require("./Rep");
const { DOM: dom, PropTypes } = React;

require("./ObjectInspector.css");

// This implements a component that renders an interactive inspector
// for looking at JavaScript objects. It expects descriptions of
// objects from the protocol, and will dynamically fetch child
// properties as objects are expanded.
//
// If you want to inspect a single object, pass the name and the
// protocol descriptor of it:
//
//  ObjectInspector({
//    name: "foo",
//    desc: { writable: true, ..., { value: { actor: "1", ... }}},
//    ...
//  })
//
// If you want multiple top-level objects (like scopes), you can pass
// an array of manually constructed nodes as `roots`:
//
//  ObjectInspector({
//    roots: [{ name: ... }, ...],
//    ...
//  });

// There are 3 types of nodes: a simple node with a children array, an
// object that has properties that should be children when they are
// fetched, and a primitive value that should be displayed with no
// children.

function nodeHasChildren(item) {
  return Array.isArray(item.contents);
}

function nodeHasProperties(item) {
  return !nodeHasChildren(item) && item.contents.value.type === "object";
}

function nodeIsPrimitive(item) {
  return !nodeHasChildren(item) && !nodeHasProperties(item);
}

function createNode(name, path, contents) {
  // The path is important to uniquely identify the item in the entire
  // tree. This helps debugging & optimizes React's rendering of large
  // lists. The path will be separated by property name,
  // i.e. `{ foo: { bar: { baz: 5 }}}` will have a path of `foo/bar/baz`
  // for the inner object.
  return { name, path, contents };
}

const ObjectInspector = React.createClass({
  propTypes: {
    name: PropTypes.string,
    desc: PropTypes.object,
    roots: PropTypes.array,
    getObjectProperties: PropTypes.func.isRequired,
    loadObjectProperties: PropTypes.func.isRequired
  },

  displayName: "ObjectInspector",

  getInitialState() {
    // Cache of dynamically built nodes. We shouldn't need to clear
    // this out ever, since we don't ever "switch out" the object
    // being inspected.
    this.actorCache = {};
    return {};
  },

  makeNodesForProperties(objProps, parentPath) {
    const { ownProperties, prototype } = objProps;

    const nodes = Object.keys(ownProperties).filter(name => {
      // Ignore non-concrete values like getters and setters
      // for now by making sure we have a value.
      return "value" in ownProperties[name];
    }).map(name => {
      return createNode(name, `${parentPath}/${name}`, ownProperties[name]);
    });

    // Add the prototype if it exists and is not null
    if (prototype && prototype.type !== "null") {
      nodes.push(createNode(
        "__proto__",
        `${parentPath}/__proto__`,
        { value: prototype }
      ));
    }

    return nodes;
  },

  getChildren(item) {
    const { getObjectProperties } = this.props;
    const obj = item.contents;

    // Nodes can either have children already, or be an object with
    // properties that we need to go and fetch.
    if (nodeHasChildren(item)) {
      return item.contents;
    } else if (nodeHasProperties(item)) {
      const actor = obj.value.actor;

      // Because we are dynamically creating the tree as the user
      // expands it (not precalcuated tree structure), we cache child
      // arrays. This not only helps performance, but is necessary
      // because the expanded state depends on instances of nodes
      // being the same across renders. If we didn't do this, each
      // node would be a new instance every render.
      const key = item.path;
      if (this.actorCache[key]) {
        return this.actorCache[key];
      }

      const loadedProps = getObjectProperties(actor);
      if (loadedProps) {
        const children = this.makeNodesForProperties(loadedProps, item.path);
        this.actorCache[actor] = children;
        return children;
      }
      return [];
    }
    return [];
  },

  renderItem(item, depth, focused, _, expanded, { setExpanded }) {
    let objectValue;
    if (nodeHasProperties(item) || nodeIsPrimitive(item)) {
      const object = item.contents.value;
      objectValue = Rep({ object, mode: "tiny" });
    }

    return dom.div(
      { className: classnames("node", { focused }),
        style: { marginLeft: depth * 15 },
        onClick: e => {
          e.stopPropagation();
          setExpanded(item, !expanded);
        }
      },
      Svg("arrow", {
        className: classnames({
          expanded: expanded,
          hidden: nodeIsPrimitive(item)
        })
      }),
      dom.span({ className: "object-label" }, item.name),
      dom.span({ className: "object-delimiter" },
               objectValue ? ": " : ""),
      dom.span({ className: "object-value" }, objectValue || "")
    );
  },

  render() {
    const { name, desc, loadObjectProperties } = this.props;

    let roots = this.props.roots;
    if (!roots) {
      roots = [createNode(name, name, desc)];
    }

    return ManagedTree({
      itemHeight: 20,
      getParent: item => null,
      getChildren: this.getChildren,
      getRoots: () => roots,
      getKey: item => item.path,
      autoExpand: 0,
      autoExpandDepth: 1,
      autoExpandAll: false,
      disabledFocus: true,
      onExpand: item => {
        if (nodeHasProperties(item)) {
          loadObjectProperties(item.contents.value);
        }
      },

      renderItem: this.renderItem
    });
  }
});

module.exports = ObjectInspector;
