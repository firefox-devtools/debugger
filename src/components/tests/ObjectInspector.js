import { shallow } from "enzyme";
import { renderItem } from "../shared/ObjectInspector";

const arrayGrip = {
  type: "object",
  class: "Array",
  preview: {
    kind: "ArrayLike",
    length: 1,
    items: [{}]
  }
};

describe("ObjectInspector", () => {
  it("should render simple values (foo: 2)", () => {
    const item = {
      name: "foo",
      path: "foo",
      contents: { value: 2 }
    };

    const wrapper = shallow(
      renderItem(item, 0, false, {}, false, {
        setExpanded: () => {}
      })
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("expands on click", () => {
    const item = {
      name: "foo",
      path: "foo",
      contents: { value: 2 }
    };

    const setExpanded = jest.fn();
    const wrapper = shallow(
      renderItem(item, 0, false, {}, false, {
        setExpanded
      })
    );
    wrapper.simulate("click", { stopPropagation: () => {} });
    expect(setExpanded).toBeCalled();
  });

  it("should render arrays ([foo: 2])", () => {
    const value = arrayGrip;
    const item = {
      name: "foo",
      path: "foo",
      contents: { value }
    };

    const wrapper = shallow(
      renderItem(item, 0, false, {}, false, {
        setExpanded: () => {}
      })
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render expanded arrays ([foo: 2])", () => {
    const value = arrayGrip;
    const item = {
      name: "foo",
      path: "foo",
      contents: { value }
    };

    const wrapper = shallow(
      renderItem(item, 0, false, {}, true, {
        setExpanded: () => {}
      })
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render functions", () => {
    const value = {
      type: "object",
      class: "Function",
      preview: {}
    };
    const item = {
      name: "foo",
      path: "foo",
      contents: { value }
    };

    const wrapper = shallow(
      renderItem(item, 0, false, {}, true, {
        setExpanded: () => {}
      })
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render optimized items", () => {
    const item = {
      name: "foo",
      path: "foo",
      contents: { value: { optimizedOut: true } }
    };

    const wrapper = shallow(
      renderItem(item, 0, false, {}, true, {
        setExpanded: () => {}
      })
    );

    const objValue = wrapper.find(".object-value").text();
    expect(wrapper).toMatchSnapshot();
    expect(objValue).toEqual("(optimized away)");
  });

  it("should render missing arguments", () => {
    const item = {
      name: "foo",
      path: "foo",
      contents: { value: { missingArguments: true } }
    };

    const wrapper = shallow(
      renderItem(item, 0, false, {}, true, {
        setExpanded: () => {}
      })
    );

    const objValue = wrapper.find(".object-value").text();
    expect(wrapper).toMatchSnapshot();
    expect(objValue).toEqual("(unavailable)");
  });
});
