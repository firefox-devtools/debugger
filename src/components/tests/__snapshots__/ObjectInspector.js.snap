// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectInspector should render arrays ([foo: 2]) 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <Rep
            defaultRep={
                Object {
                    "rep": [Function],
                    "supportsObject": [Function],
                  }
            }
            mode={Symbol(TINY)}
            object={
                Object {
                    "class": "Array",
                    "preview": Object {
                      "items": Array [
                        Object {},
                      ],
                      "kind": "ArrayLike",
                      "length": 1,
                    },
                    "type": "object",
                  }
            }
        />
    </span>
</div>,
  "nodes": Array [
    <div
      className="node object-node"
      onClick={[Function]}
      onDoubleClick={[Function]}
      style={
            Object {
                  "marginLeft": 0,
                }
      }
>
      <InlineSVG
            className="arrow"
            element="i"
            raw={false}
            src="<svg></svg>"
      />
      <span
            className="object-label"
            dir="ltr"
            onClick={[Function]}
      >
            foo
      </span>
      <span
            className="object-delimiter"
      >
            : 
      </span>
      <span
            className="object-value"
      >
            <Rep
                  defaultRep={
                        Object {
                              "rep": [Function],
                              "supportsObject": [Function],
                            }
                  }
                  mode={Symbol(TINY)}
                  object={
                        Object {
                              "class": "Array",
                              "preview": Object {
                                "items": Array [
                                  Object {},
                                ],
                                "kind": "ArrayLike",
                                "length": 1,
                              },
                              "type": "object",
                            }
                  }
            />
      </span>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": null,
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <Rep
            defaultRep={
                Object {
                    "rep": [Function],
                    "supportsObject": [Function],
                  }
            }
            mode={Symbol(TINY)}
            object={
                Object {
                    "class": "Array",
                    "preview": Object {
                      "items": Array [
                        Object {},
                      ],
                      "kind": "ArrayLike",
                      "length": 1,
                    },
                    "type": "object",
                  }
            }
        />
    </span>
</div>,
}
`;

exports[`ObjectInspector should render expanded arrays ([foo: 2]) 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow expanded"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <Rep
            defaultRep={
                Object {
                    "rep": [Function],
                    "supportsObject": [Function],
                  }
            }
            mode={Symbol(TINY)}
            object={
                Object {
                    "class": "Array",
                    "preview": Object {
                      "items": Array [
                        Object {},
                      ],
                      "kind": "ArrayLike",
                      "length": 1,
                    },
                    "type": "object",
                  }
            }
        />
    </span>
</div>,
  "nodes": Array [
    <div
      className="node object-node"
      onClick={[Function]}
      onDoubleClick={[Function]}
      style={
            Object {
                  "marginLeft": 0,
                }
      }
>
      <InlineSVG
            className="arrow expanded"
            element="i"
            raw={false}
            src="<svg></svg>"
      />
      <span
            className="object-label"
            dir="ltr"
            onClick={[Function]}
      >
            foo
      </span>
      <span
            className="object-delimiter"
      >
            : 
      </span>
      <span
            className="object-value"
      >
            <Rep
                  defaultRep={
                        Object {
                              "rep": [Function],
                              "supportsObject": [Function],
                            }
                  }
                  mode={Symbol(TINY)}
                  object={
                        Object {
                              "class": "Array",
                              "preview": Object {
                                "items": Array [
                                  Object {},
                                ],
                                "kind": "ArrayLike",
                                "length": 1,
                              },
                              "type": "object",
                            }
                  }
            />
      </span>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": null,
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow expanded"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <Rep
            defaultRep={
                Object {
                    "rep": [Function],
                    "supportsObject": [Function],
                  }
            }
            mode={Symbol(TINY)}
            object={
                Object {
                    "class": "Array",
                    "preview": Object {
                      "items": Array [
                        Object {},
                      ],
                      "kind": "ArrayLike",
                      "length": 1,
                    },
                    "type": "object",
                  }
            }
        />
    </span>
</div>,
}
`;

exports[`ObjectInspector should render functions 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow expanded"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        <span
            className="function-signature"
        >
            <span
                className="function-name"
            >
                foo
            </span>
            <span
                className="paren"
            >
                (
            </span>
            <span
                className="paren"
            >
                )
            </span>
        </span>
    </span>
    <span
        className="object-delimiter"
    />
    <span
        className="object-value"
    />
</div>,
  "nodes": Array [
    <div
      className="node object-node"
      onClick={[Function]}
      onDoubleClick={[Function]}
      style={
            Object {
                  "marginLeft": 0,
                }
      }
>
      <InlineSVG
            className="arrow expanded"
            element="i"
            raw={false}
            src="<svg></svg>"
      />
      <span
            className="object-label"
            dir="ltr"
            onClick={[Function]}
      >
            <span
                  className="function-signature"
            >
                  <span
                        className="function-name"
                  >
                        foo
                  </span>
                  <span
                        className="paren"
                  >
                        (
                  </span>
                  <span
                        className="paren"
                  >
                        )
                  </span>
            </span>
      </span>
      <span
            className="object-delimiter"
      />
      <span
            className="object-value"
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": null,
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow expanded"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        <span
            className="function-signature"
        >
            <span
                className="function-name"
            >
                foo
            </span>
            <span
                className="paren"
            >
                (
            </span>
            <span
                className="paren"
            >
                )
            </span>
        </span>
    </span>
    <span
        className="object-delimiter"
    />
    <span
        className="object-value"
    />
</div>,
}
`;

exports[`ObjectInspector should render missing arguments 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow expanded hidden"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <span
            className="unavailable"
        >
            (unavailable)
        </span>
    </span>
</div>,
  "nodes": Array [
    <div
      className="node object-node"
      onClick={[Function]}
      onDoubleClick={[Function]}
      style={
            Object {
                  "marginLeft": 0,
                }
      }
>
      <InlineSVG
            className="arrow expanded hidden"
            element="i"
            raw={false}
            src="<svg></svg>"
      />
      <span
            className="object-label"
            dir="ltr"
            onClick={[Function]}
      >
            foo
      </span>
      <span
            className="object-delimiter"
      >
            : 
      </span>
      <span
            className="object-value"
      >
            <span
                  className="unavailable"
            >
                  (unavailable)
            </span>
      </span>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": null,
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow expanded hidden"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <span
            className="unavailable"
        >
            (unavailable)
        </span>
    </span>
</div>,
}
`;

exports[`ObjectInspector should render optimized items 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow expanded hidden"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <span
            className="unavailable"
        >
            (optimized away)
        </span>
    </span>
</div>,
  "nodes": Array [
    <div
      className="node object-node"
      onClick={[Function]}
      onDoubleClick={[Function]}
      style={
            Object {
                  "marginLeft": 0,
                }
      }
>
      <InlineSVG
            className="arrow expanded hidden"
            element="i"
            raw={false}
            src="<svg></svg>"
      />
      <span
            className="object-label"
            dir="ltr"
            onClick={[Function]}
      >
            foo
      </span>
      <span
            className="object-delimiter"
      >
            : 
      </span>
      <span
            className="object-value"
      >
            <span
                  className="unavailable"
            >
                  (optimized away)
            </span>
      </span>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": null,
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow expanded hidden"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <span
            className="unavailable"
        >
            (optimized away)
        </span>
    </span>
</div>,
}
`;

exports[`ObjectInspector should render simple values (foo: 2) 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow hidden"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <Rep
            defaultRep={
                Object {
                    "rep": [Function],
                    "supportsObject": [Function],
                  }
            }
            mode={Symbol(TINY)}
            object={2}
        />
    </span>
</div>,
  "nodes": Array [
    <div
      className="node object-node"
      onClick={[Function]}
      onDoubleClick={[Function]}
      style={
            Object {
                  "marginLeft": 0,
                }
      }
>
      <InlineSVG
            className="arrow hidden"
            element="i"
            raw={false}
            src="<svg></svg>"
      />
      <span
            className="object-label"
            dir="ltr"
            onClick={[Function]}
      >
            foo
      </span>
      <span
            className="object-delimiter"
      >
            : 
      </span>
      <span
            className="object-value"
      >
            <Rep
                  defaultRep={
                        Object {
                              "rep": [Function],
                              "supportsObject": [Function],
                            }
                  }
                  mode={Symbol(TINY)}
                  object={2}
            />
      </span>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": null,
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <div
    className="node object-node"
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
        Object {
            "marginLeft": 0,
          }
    }
>
    <InlineSVG
        className="arrow hidden"
        element="i"
        raw={false}
        src="<svg></svg>"
    />
    <span
        className="object-label"
        dir="ltr"
        onClick={[Function]}
    >
        foo
    </span>
    <span
        className="object-delimiter"
    >
        : 
    </span>
    <span
        className="object-value"
    >
        <Rep
            defaultRep={
                Object {
                    "rep": [Function],
                    "supportsObject": [Function],
                  }
            }
            mode={Symbol(TINY)}
            object={2}
        />
    </span>
</div>,
}
`;
